### ISSUE TITLE
Support for PEP 695 - E0001 should not fire on 3.12 generics syntax

### ISSUE THREAD

**Issue by rbobillot** [pylint-dev/pylint#9217]
### Current problem

I am using Python 3.12 generics new syntax [PEP-695](https://docs.python.org/3/whatsnew/3.12.html#whatsnew312-pep695) in a project, such as:
```python
def get_users[I](skip: int | None, limit: int | None, inventory: I) -> list[User]:
    return repo_handler.get_users(skip, limit, inventory)
```

Unfortunately, pylint cannot parse it and the following error occurs:
```
repository.py:18:8: E0001: Parsing failed: 'invalid syntax (<unknown>, line 18)' (syntax-error)
```

I even tried with the basic syntax:
```python
def head[T](args: Iterable[T]) -> T:
    return args[0]
```
But the same error occurs

### Desired solution

For now, I'd say it should give a more ad-hoc error (rather than `invalid syntax`), so we could at least skip it.
Ideally it should handle this new syntax.

### Additional context

I didn't find this issue among `pylint` open issues.

Though, I found it [here](https://github.com/python/mypy/issues/15238) in `mypy` issues
For now, they did [a little workaround](https://github.com/python/mypy/pull/16013), saying they cant handle PEP-695 yet,
but it seems that a lot of work needs to be done to handle it.

Unfortunately I'm not sure I have enough free time yet to open a PR and try to fix those issues

**Comment by jacobtylerwalls** [comment]
Hi 👋🏻. What python interpreter are you using to run pylint? Below 3.12?

**Comment by Pierre-Sassoulas** [comment]
Hmm yeah, I reproduced on 3.11, but a 3.11 interpreter is supposed to have this result.

**Comment by rbobillot** [comment]
I use the 3.12 interpreter (via pyenv)

> Hi 👋🏻. What python interpreter are you using to run pylint? Below 3.12?

**Comment by jacobtylerwalls** [comment]
Thanks for providing the output of `pylint --version`. This is the only report we have of 3.12 interpreters not parsing the new syntax. (We've had multiple reporters report subtler issues about pylint's static analysis of these correctly parsed constructs, instead.) In other words, pylint/astroid wraps python's AST parser, so I find this behavior surprising--I'd suggest checking for non-obvious things, like pylint being invoked differently between runs, or differences between your local setup and CI.

From your first error message, it also seems pylint is being executed on an interpreter on 3.10 or below, because in 3.11 I would expect a more fine-grained error message (`expected '('`):

```py
% python3.12 -m pylint b.py
************* Module b
b.py:1:0: C0114: Missing module docstring (missing-module-docstring)
b.py:1:0: C0116: Missing function or method docstring (missing-function-docstring)
b.py:1:76: E0602: Undefined variable 'User' (undefined-variable)
b.py:1:14: C0103: Variable name "I" doesn't conform to snake_case naming style (invalid-name)
b.py:2:11: E0602: Undefined variable 'repo_handler' (undefined-variable)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)

% python3.11 -m pylint b.py
************* Module b
b.py:1:14: E0001: Parsing failed: 'expected '(' (<unknown>, line 1)' (syntax-error)

% python3.10 -m pylint b.py
************* Module b
b.py:1:14: E0001: Parsing failed: 'invalid syntax (<unknown>, line 1)' (syntax-error)
```

**Comment by DJawna** [comment]
i have got the same problem and on my system i only have the newest version of python
